README.rst

This plugin installs and runs gmconvert for converting structures and volumes to Gaussian mixture models.

**Install the plugin in devel mode**

.. code-block::

    scipion3 installp -p /home/me/scipion-em-gmconvert --devel

If installation fails, you can access pip options like:

.. code-block::

    scipion3 python -m pip ... (list, install, uninstall)


CHANGES.txt
5-June-2023
    * First complete plugin with 2 basic protocols

gmconvert/tests/__init__.py
# **************************************************************************
# *
# * Authors:     James Krieger (jmkrieger@cnb.csic.es)
# *
# * Unidad de  Bioinformatica of Centro Nacional de Biotecnologia , CSIC
# *
# * This program is free software; you can redistribute it and/or modify
# * it under the terms of the GNU General Public License as published by
# * the Free Software Foundation; either version 3 of the License, or
# * (at your option) any later version.
# *
# * This program is distributed in the hope that it will be useful,
# * but WITHOUT ANY WARRANTY; without even the implied warranty of
# * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# * GNU General Public License for more details.
# *
# * You should have received a copy of the GNU General Public License
# * along with this program; if not, write to the Free Software
# * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
# * 02111-1307  USA
# *
# *  All comments concerning this program package may be sent to the
# *  e-mail address 'scipion@cnb.csic.es'
# *
# **************************************************************************

from .test_protocols_gmconvert import TestGMConvert


gmconvert/tests/test_protocols_gmconvert.py
# **************************************************************************
# *
# * Authors:     James Krieger (jmkrieger@cnb.csic.es)
# *
# * Biocomputing Unit, Centro Nacional de Biotecnologia, CSIC
# *
# * This program is free software; you can redistribute it and/or modify
# * it under the terms of the GNU General Public License as published by
# * the Free Software Foundation; either version 3 of the License, or
# * (at your option) any later version.
# *
# * This program is distributed in the hope that it will be useful,
# * but WITHOUT ANY WARRANTY; without even the implied warranty of
# * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# * GNU General Public License for more details.
# *
# * You should have received a copy of the GNU General Public License
# * along with this program; if not, write to the Free Software
# * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
# * 02111-1307  USA
# *
# *  All comments concerning this program package may be sent to the
# *  e-mail address 'scipion@cnb.csic.es'
# *
# **************************************************************************

from pyworkflow.tests import BaseTest, setupTestProject
from pwem.protocols import ProtImportVolumes, ProtImportPdb
import pwem
import shutil, os

from ..protocols import GMConvertAtomStruct, GMConvertVolume


FROM_FILE, FROM_SCIPION = 0, 1

class TestGMConvert(BaseTest):
    @classmethod
    def setUpClass(cls):
        setupTestProject(cls)

    def test_gmconvertPdb(cls):
        cls._runImportPDB()
        cls._runConvertPdb()

    def test_gmconvertVol(cls):
        cls._runImportVolumes()
        cls._runConvertVol()

    @classmethod
    def _runImportVolumes(cls):
        """ Run an Import volumes protocol. """
        protImportVol = cls.newProtocol(
            ProtImportVolumes,
            importFrom=1,
            emdbId='2190')
        cls.launchProtocol(protImportVol)
        cls.protImportVol = protImportVol

    @classmethod
    def _runImportPDB(cls):
        protImportPDB = cls.newProtocol(
            ProtImportPdb,
            inputPdbData=0,
            pdbId='5c44')
        cls.launchProtocol(protImportPDB)
        cls.protImportPDB = protImportPDB

    def _runConvertVol(self):
        protVol = self.newProtocol(
            GMConvertVolume,
            inputVolume=self.protImportVol.outputVolume,
            cutoff=0.0666, numGaussians=10,
            outFn='2190_ng10.gmm')

        self.launchProtocol(protVol)
        gmmOut = getattr(protVol, 'outFile', None)
        self.assertIsNotNone(gmmOut)

    def _runConvertPdb(self):
        protPdb = self.newProtocol(
            GMConvertAtomStruct,
            inputStructure=self.protImportPDB.outputPdb,
            numGaussians=10,
            outFn='5c44_ng10.gmm')

        self.launchProtocol(protPdb)
        gmmOut = getattr(protPdb, 'outFile', None)
        self.assertIsNotNone(gmmOut)





gmconvert/bibtex.py
# -*- coding: utf-8 -*-
# **************************************************************************
# *
# * Authors:     you (you@yourinstitution.email)
# *
# * your institution
# *
# * This program is free software; you can redistribute it and/or modify
# * it under the terms of the GNU General Public License as published by
# * the Free Software Foundation; either version 2 of the License, or
# * (at your option) any later version.
# *
# * This program is distributed in the hope that it will be useful,
# * but WITHOUT ANY WARRANTY; without even the implied warranty of
# * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# * GNU General Public License for more details.
# *
# * You should have received a copy of the GNU General Public License
# * along with this program; if not, write to the Free Software
# * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
# * 02111-1307  USA
# *
# *  All comments concerning this program package may be sent to the
# *  e-mail address 'scipion@cnb.csic.es'
# *
# **************************************************************************

"""
@article{Kawabata2018,
title = Gaussian-input Gaussian mixture model for representing density maps and atomic models",
journal = "J Struct Biol.",
volume = "203",
number = "1",
pages = "1 - 16",
year = "2018",
issn = "0000-1111",
doi = "http://dx.doi.org/10.1016/j.jsb.2018.03.002",
url = "http://www.url",
author = "Kawabata, T.",
keywords = "Gaussian mixture model fitting",
}
"""


gmconvert/__init__.py
# **************************************************************************
# *
# * Authors:     James Krieger (jmkrieger@cnb.csic.es)
# *
# * Biocomputing Unit, Centro Nacional de Biotecnologia, CSIC
# *
# * This program is free software; you can redistribute it and/or modify
# * it under the terms of the GNU General Public License as published by
# * the Free Software Foundation; either version 2 of the License, or
# * (at your option) any later version.
# *
# * This program is distributed in the hope that it will be useful,
# * but WITHOUT ANY WARRANTY; without even the implied warranty of
# * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# * GNU General Public License for more details.
# *
# * You should have received a copy of the GNU General Public License
# * along with this program; if not, write to the Free Software
# * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
# * 02111-1307  USA
# *
# *  All comments concerning this program package may be sent to the
# *  e-mail address 'scipion@cnb.csic.es'
# *
# **************************************************************************

import pwem
from .constants import *
from os.path import join
import subprocess

_logo = "icon.png"
_references = ['Kawabata2018']

GMCONVERT_DIC = {'name': 'gmconvert', 'version': '20220509', 'home': 'GMCONVERT_HOME'}

class Plugin(pwem.Plugin):
    _homeVar = GMCONVERT_DIC['home']
    _pathVars = [GMCONVERT_DIC['home']]
    _supportedVersions = [LATEST]
    _gmconvertName = GMCONVERT_DIC['name'] + '-' + GMCONVERT_DIC['version']

    @classmethod
    def _defineVariables(cls):
        """ Return and write a variable in the config file.
        """
        cls._defineEmVar(GMCONVERT_DIC['home'], cls._gmconvertName)

    @classmethod
    def defineBinaries(cls, env):
        installationCmd = 'wget -O gmconvert-{}.tar.gz {} --no-check-certificate && '. \
          format(GMCONVERT_DIC['version'], cls._getGMConvertDownloadUrl())
        installationCmd += 'tar -xf gmconvert-{}.tar.gz && '.\
          format(GMCONVERT_DIC['version'])
        installationCmd += 'cd src && make &&'

        # Creating validation file
        GMCONVERT_INSTALLED = '%s_installed' % GMCONVERT_DIC['name']
        installationCmd += 'cd .. && touch %s' % GMCONVERT_INSTALLED  # Flag installation finished

        env.addPackage(GMCONVERT_DIC['name'],
                       version=GMCONVERT_DIC['version'],
                       tar='void.tgz',
                       commands=[(installationCmd, GMCONVERT_INSTALLED)],
                       default=True)

    @classmethod
    def _getGMConvertDownloadUrl(cls):
        return 'https://pdbj.org/gmfit/cgi-bin/dwnld_src_file.cgi?filename=gmconvert-src-{}.tar.gz'.format(GMCONVERT_DIC['version'])

    @classmethod
    def getGMConvertBin(cls):
        return join(cls.getVar(GMCONVERT_DIC['home']), 'gmconvert')

    @classmethod
    def runGMConvert(cls, protocol, args='', cwd=None):
        """ Run Gromacs command from a given protocol. """
        protocol.runJob(cls.getGMConvertBin(), args, cwd=cwd)


gmconvert/constants.py
# **************************************************************************
# *
# * Authors:     James Krieger (jmkrieger@cnb.csic.es)
# *
# * Centro Nacional de Biotecnologia, CSIC
# *
# * This program is free software; you can redistribute it and/or modify
# * it under the terms of the GNU General Public License as published by
# * the Free Software Foundation; either version 3 of the License, or
# * (at your option) any later version.
# *
# * This program is distributed in the hope that it will be useful,
# * but WITHOUT ANY WARRANTY; without even the implied warranty of
# * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# * GNU General Public License for more details.
# *
# * You should have received a copy of the GNU General Public License
# * along with this program; if not, write to the Free Software
# * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
# * 02111-1307  USA
# *
# *  All comments concerning this program package may be sent to the
# *  e-mail address 'scipion@cnb.csic.es'
# *
# **************************************************************************


LATEST = '20220509'
VERSIONS = [LATEST]
GMCONVERT_DEFAULT_VER_NUM = LATEST


gmconvert/protocols/protocol_gmconvert_structure.py
# -*- coding: utf-8 -*-
# **************************************************************************
# *
# * Authors:     James Krieger (jmkrieger@cnb.csic.es)
# *
# * Biocomputing Unit, Centro Nacional de Biotecnologia, CSIC
# *
# * This program is free software; you can redistribute it and/or modify
# * it under the terms of the GNU General Public License as published by
# * the Free Software Foundation; either version 2 of the License, or
# * (at your option) any later version.
# *
# * This program is distributed in the hope that it will be useful,
# * but WITHOUT ANY WARRANTY; without even the implied warranty of
# * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# * GNU General Public License for more details.
# *
# * You should have received a copy of the GNU General Public License
# * along with this program; if not, write to the Free Software
# * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
# * 02111-1307  USA
# *
# *  All comments concerning this program package may be sent to the
# *  e-mail address 'jmkrieger@cnb.csic.es'
# *
# **************************************************************************


"""
This module will provide conversion of an atomic structure to a Gaussian mixture model
"""
from pyworkflow.protocol import Protocol, params
from pwem.objects import EMFile

from gmconvert import Plugin as gmconvertPlugin

class GMConvertAtomStruct(Protocol):
    """
    This protocol will convert an atomic structure to a Gaussian mixture model
    """
    _label = 'gmconvert atomic structure'

    # -------------------------- DEFINE param functions ----------------------
    def _defineParams(self, form):
        """ Define the input parameters that will be used.
        Params:
            form: this is the form to be populated with sections and params.
        """
        # You need a params to belong to a section:
        form.addSection(label='gmconvert atomic structure')
        form.addParam('inputStructure', params.PointerParam, label="Input structure",
                      important=True,
                      pointerClass='AtomStruct',
                      help='The input structure can be an atomic model '
                           '(true PDB) or a pseudoatomic model '
                           '(an EM volume converted into pseudoatoms)')

        form.addParam('numGaussians', params.IntParam,
                      default=10, label='number of Gaussians', important=True)

        form.addParam('outFn', params.StringParam,
                      default='', label='Output filename', important=True)

    # --------------------------- STEPS functions ------------------------------
    def _insertAllSteps(self):
        # Insert processing steps
        self._insertFunctionStep('convertStep')
        self._insertFunctionStep('createOutputStep')

    def convertStep(self):
        args = 'A2G -ipdb {0} -ng {1} -ogmm {2}'.format(self.inputStructure.get().getFileName(),
                                                        self.numGaussians.get(), 
                                                        self._getPath(self.outFn.get()))
        try:
            gmconvertPlugin.runGMConvert(self, args)
        except:
            # check if it actually finished correctly
            fi = open(self._getLogsPath('run.stdout'), 'r')
            lines = fi.readlines()
            fi.close()

            if not lines[-1].startswith('COMP_TIME_SEC_FINAL'):
                # gmconvert usually raises an non-zero error code regardless 
                # so we should go and read the error information from the log file
                log = open(self._getLogsPath("run.stderr"), 'r')
                if len(log.read().splitlines()) > 0:
                    for line in log.read().splitlines():
                        self.info("ERROR: %s." % line)
                        raise ChildProcessError("gmconvert has failed: %s. See error log "
                                                "for more details." % line) from None


    def createOutputStep(self):
        # register how many times the message has been printed
        # Now count will be an accumulated value
        outputFile = EMFile(filename=self._getPath(self.outFn.get()))
        self._defineOutputs(outFile=outputFile)

    # --------------------------- INFO functions -----------------------------------
    def _summary(self):
        """ Summarize what the protocol has done"""
        summary = []
        if self.isFinished():
            summary.append("This protocol has converted {0} to a GMM in {0}." % (self.inputStructure.get().getFileName(), 
                                                                                 self._getPath(self.outFn.get())))
        return summary

    def _methods(self):
        methods = []
        if self.isFinished():
            methods.append("This protocol has converted {0} to a GMM in {0}." % (self.inputStructure.get().getFileName(), 
                                                                                 self._getPath(self.outFn.get())))
        return methods


gmconvert/protocols/protocol_gmconvert_volume.py
# -*- coding: utf-8 -*-
# **************************************************************************
# *
# * Authors:     James Krieger (jmkrieger@cnb.csic.es)
# *
# * Biocomputing Unit, Centro Nacional de Biotecnologia, CSIC
# *
# * This program is free software; you can redistribute it and/or modify
# * it under the terms of the GNU General Public License as published by
# * the Free Software Foundation; either version 2 of the License, or
# * (at your option) any later version.
# *
# * This program is distributed in the hope that it will be useful,
# * but WITHOUT ANY WARRANTY; without even the implied warranty of
# * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# * GNU General Public License for more details.
# *
# * You should have received a copy of the GNU General Public License
# * along with this program; if not, write to the Free Software
# * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
# * 02111-1307  USA
# *
# *  All comments concerning this program package may be sent to the
# *  e-mail address 'jmkrieger@cnb.csic.es'
# *
# **************************************************************************


"""
This module will provide the conversion of a volume to a Gaussian mixture model
"""
from pyworkflow.protocol import Protocol, params
from pwem.objects import EMFile

from gmconvert import Plugin as gmconvertPlugin

class GMConvertVolume(Protocol):
    """
    This protocol will convert a volume to a Gaussian mixture model
    """
    _label = 'gmconvert volume'

    # -------------------------- DEFINE param functions ----------------------
    def _defineParams(self, form):
        """ Define the input parameters that will be used.
        Params:
            form: this is the form to be populated with sections and params.
        """
        # You need a params to belong to a section:
        form.addSection(label='gmconvert volume')
        form.addParam('inputVolume', params.PointerParam, label="Input volume",
                      important=True,
                      pointerClass='Volume',
                      help='The volume to be converted')
        
        form.addParam('cutoff', params.FloatParam,
                      default=0.05, label='threshold', important=True,
                      help='cutoff for thresholding the volume')        

        form.addParam('numGaussians', params.IntParam,
                      default=10, label='number of Gaussians', important=True)

        form.addParam('outFn', params.StringParam,
                      default='', label='Output filename', important=True)

    # --------------------------- STEPS functions ------------------------------
    def _insertAllSteps(self):
        # Insert processing steps
        self._insertFunctionStep('convertStep')
        self._insertFunctionStep('createOutputStep')

    def convertStep(self):
        args = 'V2G -imap {0} -ogmm {1} -cutoff {2} -ng {3}'.format(self.inputVolume.get().getFileName(), 
                                                                    self._getPath(self.outFn.get()), 
                                                                    self.cutoff.get(), 
                                                                    self.numGaussians.get())
        try:
            gmconvertPlugin.runGMConvert(self, args)
        except:
            # check if it actually finished correctly
            fi = open(self._getLogsPath('run.stdout'), 'r')
            lines = fi.readlines()
            fi.close()

            if not lines[-1].startswith('COMP_TIME_SEC_FINAL'):
                # gmconvert usually raises an non-zero error code regardless 
                # so we should go and read the error information from the log file
                log = open(self._getLogsPath("run.stderr"), 'r')
                if len(log.read().splitlines()) > 0:
                    for line in log.read().splitlines():
                        self.info("ERROR: %s." % line)
                        raise ChildProcessError("gmconvert has failed: %s. See error log "
                                                "for more details." % line) from None

    def createOutputStep(self):
        # register how many times the message has been printed
        # Now count will be an accumulated value
        outputFile = EMFile(filename=self._getPath(self.outFn.get()))
        self._defineOutputs(outFile=outputFile)

    # --------------------------- INFO functions -----------------------------------
    def _summary(self):
        """ Summarize what the protocol has done"""
        summary = []
        if self.isFinished():
            summary.append("This protocol has converted {0} to a GMM in {0}." % (self.inputStructure.get().getFileName(), 
                                                                                 self._getPath(self.outFn.get())))
        return summary

    def _methods(self):
        methods = []
        if self.isFinished():
            methods.append("This protocol has converted {0} to a GMM in {0}." % (self.inputStructure.get().getFileName(), 
                                                                                 self._getPath(self.outFn.get())))
        return methods
